netket/hilbert/spin.py

```diff
@@ -83,10 +82,10 @@ def random_state(self, size=None, *, out=None, rgen=None):
         shape = (*size, self._size) if size is not None else (self._size,)
 
         if out is None:
-            out = _np.empty(shape=shape)
+            out = np.empty(shape=shape)
 
         if rgen is None:
-            rgen = _random
+            rgen = np.random.default_rng()
```

Comment:
I think we were using our own random generator to make sure each MPI node had a different seed?

Reply:
Since in this PR we are goin all-in on using jax (and flax), I think it would be best if we also make the random number interface consistent with what jax is doing.

To this end, I'd propose we 1) have all function using randomness take a key (obtained from from `jax.random.PRNGKey(seed)` and `jax.random.split(key)` as appropriate) as input and 2) add a utility function `k = nk.util.mpi_split_key(key)` which takes a key on the root node (or could be a seed) and returns a different derived key on each MPI node (but deterministically based on the input). This could be implemented simply as
```python
def mpi_split_key(in_key, comm=MPI.COMM_WORLD):
    # Maybe add error/warning if in_key is not the same
    # on all MPI nodes?
    keys = jax.random.split(in_key, num=comm.size)
    return comm.bcast(keys, root=0)[comm.rank]
    # If in_key is the same on all nodes, this is equivalent:
    return keys[comm.rank]
```
and then one can explicitly use this key when needed in NetKet, like:
```python
key = nk.util.mpi_split_key(jax.random.PRNGKey(seed))
# key is now different on each MPI node, but still deterministically depends on seed
# could also be combined into one function, like `key = nk.util.init_random(seed)`
k1, k2 = jax.random.split(key)
...
ma.init_random_pars(key=k1) # probably only the root key should be used here anyways?
sampler = nk.sampler.MetropolisLocal(..., key=k2) # different sampler state for each MPI node
```

Then we can freely make use of `jax.random` internally and we have a very straightforward way of ensuring that everything deterministically depends on the seed for reproducibility (which I think is a huge plus of the JAX interface - it's hard to accidentally introduce non-deterministic randomness).

@PhilipVinc, @gcarleo What do you think?

Reply:
I completely agree and actually was already writing the same function, so... 👍 

However, the issue here is that `random_state` is a legacy function (that I did not remove) which dose not use `jax`.

The new samplers do not use this function, rather they use the random_state function defined in `netket.sampler.hilbert`, which follows jax conventions. 
I left it simply as a utility for users.

But since having two parts of the code doing the same thing is not a great idea, I propose to:
 - remove the current `hilbert.random_state`
 - replace it with `random_state(key, size, dtype)` functions defined in `netket.sampler.hilbert`, 

As it is sometimes useful to generate some random states for playing in the REPL, we could also make the key optional..?

What do you think?

Reply:
I think replacing the current random state function in Hilbert by your `random_state(key, size, dtype)` is indeed the way to go.

Regarding the optional key: Personally, I'd favor consistency over the slight convenience here - even in the REPL. In order to reduce the need for too many cumbersome calls to `jax.random.split`, we could add the `PRNGSequence` type from haiku to NetKet (since haiku is Apache-licensed, this should not be a problem):
https://github.com/deepmind/dm-haiku/blob/14f096026f9f1f792d90d8e3da7eb8c4fbace4c7/haiku/_src/base.py#L491-L504

Then, in both REPL and scripts, one can just do
```python
>>> rseq = netket.util.PRNGSequence(42)
>>> hi.random_state(next(rseq), 1)
DeviceArray([[ 1., -1., -1.,  1.]], dtype=float32)
>>> hi.random_state(next(rseq), 1)
DeviceArray([[-1., -1.,  1.,  1.]], dtype=float32)
>>> sampler = MetropolisLocal(..., key=next(rseq))
```
I think this is sufficiently convenient and the reduced room for accidental non-determinism is well worth the few additional characters to type `next(rseq)` from time to time.

We could even extend it a bit to make it MPI-aware, so that
```python
rseq = netket.util.PRNGSequence(42, mpi=True)
```
gives you a different sequence on each node.

Reply:
(I added the `mpi_split_key` with the latest commits and use it to initialise the samplers. We still have to finalise the random_state, though).

I don't mind adding a PRNGSequence.  
By the way, even if numba does not support it yet, numpy has deprecated the old 'global' rng in favour of generators, much like this you are showing.

Still I hate that one can't quickly generate some random states without also writing a lot of code when playing with the REPL to explore or debug...

Reply:
We could add a global `PRNGSequence` and default to that, like
```python
# netket/random.py
_rng = PRNGSequence(0, mpi=True)
def global_init_random(seed: int):
    global _rng
    _rng = PRNGSequence(seed, mpi=True)

def global_random_key():
    return next(_rng)
```
```python
def random_state(..., key=None):
    if key is None:
        key = netket.random.global_random_key()
```